### ISSUE TITLE
bug: 

### ISSUE THREAD

**Issue by lincolnthree** [ionic-team/ionic#21607]
# Bug Report

**Ionic version:**

[ ] **4.x**
[x] **5.x**

**Current behavior:**
Button handlers in ActionSheets do not allow user-gestures to successfully invoke Share, Clipboard or other Web APIs in PWA mode on iOS/Safari.

The browser does not appear to think that the event occurred because of a user gesture, and results in: `NotAllowedError: The request is not allowed by the user agent or the platform in the current context, possibly because the user denied permission.`

**Expected behavior:**
The handler should execute in the same promise chain as the user gesture, so that native APIs and browser functionality can be used. Safari is apparently more strict about this.

**Steps to reproduce:**
Use the following code in an ActionSheet handler in an installed PWA on iOS:

```
    return navigator.share({
      text: 'How to implement the Web Share API and a fallback',
      url: 'https://www.apple.com'
    });
```

The same code works fine when executed directly from a button click in the page template, or other source.

**Related code:**

```
    const actionSheet = await this.actionCtrl.create({
      buttons: [{
      text: 'Share',
      icon: 'share',
      handler: () => {
         return navigator.share({
           text: 'How to implement the Web Share API and a fallback',
           url: 'https://www.apple.com'
         });
      }
    }]
    });
    await actionSheet.present();
```

(Any chance we could get an updated stackblitz template for ionic 5? The above is for Ionic 4 :) )

```
insert short code snippets here
```

**Other information:**
Tested on several iOS devices/versions. All tested behaved the same way. 

**Ionic info:**

```
sharktop:topdecked-unified lincoln$ ionic info

Ionic:

   Ionic CLI                     : 6.9.2 (/usr/local/lib/node_modules/@ionic/cli)
   Ionic Framework               : @ionic/angular 5.2.2
   @angular-devkit/build-angular : 0.1000.0-rc.5
   @angular-devkit/schematics    : 10.0.0-rc.5
   @angular/cli                  : 10.0.0-rc.5
   @ionic/angular-toolkit        : 2.2.0

Cordova:

   Cordova CLI       : 8.1.2 (cordova-lib@8.1.1)
   Cordova Platforms : not available
   Cordova Plugins   : not available

Utility:

   cordova-res : not installed
   native-run  : not installed

System:

   Android SDK Tools : 26.1.1 (/Users/lincoln/Library/Android/sdk)
   NodeJS            : v12.13.0 (/usr/local/lib/node_modules/node/bin/node)
   npm               : 6.14.4
   OS                : macOS Catalina
   Xcode             : Xcode 11.5 Build version 11E608c

```

**Comment by lincolnthree** [comment]
@liamdebeasi Sorry to bug you directly. Any ideas on how I could dig in to this or where to look? Were you able to reproduce?

I'm not really super familiar with how browsers "decide" if something came from a user interaction. Happy to try to chase this down.

**Comment by liamdebeasi** [comment]
I was able to reproduce the issue. One of the things we want to look at with Framework v6 is the API for overlay components like action sheets. I think this (and many other issues) could be solved if users could provide a template directly and bind their own event handlers.

It looks like this issue occurs because the framework manually calls the button handlers instead of binding click events directly: https://github.com/ionic-team/ionic/blob/master/core/src/components/action-sheet/action-sheet.tsx#L167

**Comment by lincolnthree** [comment]
Hmm... interesting. That sounds like a significant change of direction for how action-sheets work. I'm not sure I'm able to envision what that would look like. I guess you'd pass in a `TemplateRef` instance for... each menu item?

I got as far as `safeCall` when I looked at it earlier. The thing the confused me was that when I tried creating my own promise chains (and using async/await) in a button (click) action, it always worked just fine, no matter how far down I got. I did not try invoking a function reference passed from somewhere else, like `safeCall` seems to do..

This seems like something: https://github.com/ionic-team/ionic/blob/master/core/src/utils/overlays.ts#L259

Is it possible that the zoneGate is somehow breaking the chain? Is it known what the actual disconnect is? Feel free to stop me if this is 10000 questions / too many. I'm curious.

**Comment by lincolnthree** [comment]
Oh, I see. I'm tired. 

The framework calls (b.handler) which is passed from the template. Right. So that's where you're saying the disconnect happens. Hmm... I'll do some testing around that.

I would think that if the handler were defined as an arrow function it would work, but... I guess that's not enough. I just love Safari.

**Comment by liamdebeasi** [comment]
Yeah, Safari/WebKit has a restriction where the handler needs to directly tied a click event/native browser event generated by the user in order for these APIs to work.

The v6 changes are not set in stone, so things may change. I would imagine the way you use it looks the same. Additionally, this would let Angular users declare the components in their template similar to how Ionic React users currently do.

**Comment by lincolnthree** [comment]
Interesting. So if you pass the function reference, it essentially breaks the chain, is what I'm understanding. Seems strange given It's all promises under the hood. I'd expect the browser would be able to trace it back. Restriction or bug?

Anyway... looking forward to V6 I guess :) Will have to think about our UI and how to work around this limitation. Adding buttons to the page is less ideal -- at least it should work though.

Actually, couldn't you dynamically create a template with the action handlers bound to it directly with the way things are now? It would be a bit extra code for you guys I guess, but if you're thinking about re-doing it all anyway...

I guess the issue with that is error handling, but... I'm not sure how that would be different with what you're describing anyway?

**Comment by lesrpo** [comment]
Yes, this happens to me if I add window.open(link, '_system') in the onDidDismiss,
